feat(ui): Save last route when going to settings (APP-181)

billyvg opened the pull request:
... this allows user to go back to the last route they were at within
the new settings.

* Add "lastProject" to latestContextStore
* Add "lastRoute" to latestContextStore
* Save last route when `OrganizationRoot` or `ProjectContext` unmounts
* Unset active project when `OrganizationRoot` mounts

lynnagara commented on src/sentry/static/sentry/app/views/projects/projectContext.jsx:
```diff
@@ -111,6 +112,11 @@ const ProjectContext = createReactClass({
 
+  componentWillUnmount() {
+    // Save last route so that we can jump back to view from settings
+    setLastRoute(this.props.location.pathname);
```
If it's a project settings page, perhaps we should skip this

billyvg replied:
Ah, good catch, I think the easiest way will probably be to introduce a ProjectRoot that wraps our project pages.

lynnagara commented on src/sentry/static/sentry/app/views/organizationRoot.jsx:
```diff
@@ -0,0 +1,28 @@
+ */
+class OrganizationRoot extends React.Component {
+  componentDidMount() {
+    setActiveProject(null);
```
Do we also want to clear last route here?

billyvg replied:
In the current state, I don't think it has an affect either way. I think we should keep `lastRoute`, since it will still be correct when this mounts (and if any of these root views every need lastRoute, it'd be accurate).

billyvg commented:
@lynnagara Summary of changes:

* I moved `setLastRoute` from `ProjectContext` to `ProjectDetailsLayout` since settings routes do not use the latter. 
* In `SettingsWrapper` I added some logic to preserve the state of `LatestContextStore` when it is mounted. This is used to determine behavior of Back Button
* Back Button tries to be smart if you load settings page directly -- it will use latest org/project from store. If you load the Settings Index page directly (i.e. nothing in context), it just chooses an org until you navigate to an org or project settings page, in which case it will use that.

lynnagara commented:
If you load the settings page directly, how about just going back to the entity based on the current URL? Currently it's always picking the project if one happens to have been loaded in the LatestContextStore, so navigating between org and project settings pages doesn't behave in a consistent way.
